ee/clickhouse/queries/clickhouse_retention.py

```diff
@@ -58,62 +50,103 @@ def calculate_retention(self, filter: Filter, team: Team, total_intervals: int)
             target_query = "AND e.event = %(target_event)s"
             target_params = {"target_event": target_entity.id}
 
-        result = sync_execute(
-            RETENTION_SQL.format(
-                target_query=target_query,
-                filters="{filters}".format(filters=prop_filters) if filter.properties else "",
-                trunc_func=trunc_func,
-            ),
-            {
-                "team_id": team.pk,
-                "start_date": date_from.strftime(
+        trunc_func = self._get_trunc_func(period)
+
+        if period == "Week":
+            date_from = date_from - timedelta(days=date_from.isoweekday() % 7)
+
+        person_q = SqlQuery(
+            SELECT="""
+                person_id,
+                distinct_id
+            """,
+            FROM="""
+                person_distinct_id
+            """,
+            WHERE="""
+                team_id = %(team_id)s
+            """,
+        ).update_params(team_id=team.pk)
+
+        event_query = compose(
+            lambda: SqlQuery(
+                SELECT="""
+                        timestamp AS event_date,
+                        pdi.person_id as person_id
+                        """.format(
+                    trunc_func=trunc_func
+                )
+            )
+            .c(
+                WHERE="""
+                        toDateTime(e.timestamp) >= toDateTime(%(start_date)s)
+                        AND toDateTime(e.timestamp) <= toDateTime(%(end_date)s)
+                        AND e.team_id = %(team_id)s
+                        """
+            )
+            .update_params(
+                start_date=date_from.strftime(
                     "%Y-%m-%d{}".format(" %H:%M:%S" if filter.period == "Hour" else " 00:00:00")
                 ),
-                "end_date": date_to.strftime(
+                end_date=date_to.strftime("%Y-%m-%d{}".format(" %H:%M:%S" if filter.period == "Hour" else " 00:00:00")),
+                team_id=team.pk,
+            )
+            .c(FROM=f"events e JOIN ({person_q}) pdi on e.distinct_id = pdi.distinct_id")
+            .update_params(**person_q.params),
+            lambda q: target_query and q.c(WHERE=target_query).update_params(**target_params),
```

Comment:
I think extracting a function + the code above dealing with target_query/target_params makes sense:

```pythondef think_of_a_name(target_entity):
    if target_entity.type == TREND_FILTER_TYPE_ACTIONS:
        action = Action.objects.get(pk=target_entity.id)
        action_query, target_params = format_action_filter(action, use_loop=True)
        return lambda q: q.c(WHERE="AND e.uuid IN ({})".format(action_query)).update_params(**target_params)
    elif target_entity.type == TREND_FILTER_TYPE_EVENTS:
        return lambda q: q.c(WHERE="AND e.event = %(target_event)s").update_params(target_event=target_entity.id)
```

compose could weed out the None case itself.

This could be beautified further with a decorator, e.g.

```python
def wrap_query(fn):
    return lambda *args: lambda q: fn(*args, q)

@wrap_query
def think_of_a_name(target_entity, query):
    if target_entity.type == TREND_FILTER_TYPE_ACTIONS:
        action = Action.objects.get(pk=target_entity.id)
        action_query, target_params = format_action_filter(action, use_loop=True)
        return q.c(WHERE="AND e.uuid IN ({})".format(action_query)).update_params(**target_params)
    elif target_entity.type == TREND_FILTER_TYPE_EVENTS:
        return q.c(WHERE="AND e.event = %(target_event)s").update_params(target_event=target_entity.id)
```

Which might be overkill